examples/diagram.dx

```diff
@@ -0,0 +1,193 @@
+def listLength ((AsList n _):List a) : Int = n
+
+-- TODO: we want this for any monoid but this implementation won't work.
+def concat (lists:n=>(List a)) : List a =
+  totalSize = sum for i. listLength lists.i
+  AsList _ $ fst $ withState 0 \listIdx.
+    fst $ withState 0 \eltIdx.
+      for i:(Fin totalSize).
+        while (\(). get eltIdx >= listLength (lists.((get listIdx)@_))) \().
+          eltIdx := 0
+          listIdx := get listIdx + 1
+        (AsList _ xs) = lists.((get listIdx)@_)
+        eltIdxVal = get eltIdx
+        eltIdx := eltIdxVal + 1
+        xs.(eltIdxVal@_)
+
+
+Point : Type = (Float & Float)
+
+data Geom =
+  Spot
+  Circle Float
+  Rectangle Float Float  -- width, height
+  Line Point
+
+-- HTML color string. TODO: consider an RGB triple instead
+Color : Type = String
+
+-- TODO: we should add overloaded string literals so we don't need this
+def str (n:Int) ?-> (s:(Fin n=>Char)) : String = AsList _ s
+
+black : Color = str "black"
+white : Color = str "white"
+gray  : Color = str "gray"
+red   : Color = str "red"
+green : Color = str "green"
+blue  : Color = str "blue"
+
+Attributes : Type = { fill   : Maybe Color
+                    & stroke : Maybe Color
+                    & lw     : Int }
+defaultAttributes : Attributes = { fill   = Nothing
+                                 , stroke = Just black
+                                 , lw = 1 }
+
+-- TODO: consider sharing attributes among a set of objects for efficiency
+data Diagram = MkDiagram (List (Attributes & Point & Geom))
+
+instance monoidDiagram : Monoid Diagram where
+  mempty = MkDiagram mempty
+  mcombine = \(MkDiagram d1) (MkDiagram d2). MkDiagram $ d1 <> d2
+
+def concatDiagrams (diagrams:n=>Diagram) : Diagram =
+  MkDiagram $ concat for i.
+    (MkDiagram d) = diagrams.i
+    d
+
+-- TODO: arbitrary affine transformations. Our current representation of
+-- rectangles and circles means we can only do scale/flip/rotate90.
+-- Should we use lenses/isomorphisms for these instead?
+def applyTransformation
+      (transformPoint: Point -> Point)
+      (transformGeom:  Geom -> Geom)
+      (d:Diagram) : Diagram =
+  (MkDiagram (AsList _ objs)) = d
+  (MkDiagram $ AsList _ for i.
+    (attr, p, geom) = objs.i
+    (attr, transformPoint p, transformGeom geom))
+
+flipY : Diagram -> Diagram =
+  applyTransformation (\(x,y). (x, -y)) \geom. case geom of
+    Spot          -> Spot
+    Circle r      -> Circle r
+    Rectangle w h -> Rectangle w h
+    Line (x, y)   -> Line (x, -y)
```

Comment:
can't we do
```suggestion
    Line (x, y)   -> Line (x, -y)
    other -> other
```
Or does pattern matchign not support that yet?

Reply:
Right, it's the same issue as #307.